panoramapublic/src/org/labkey/panoramapublic/PanoramaPublicController.java

```diff
@@ -705,6 +716,176 @@ public Container getSupportContainer()
+            String doiPrefix = form.getDoiPrefix();
+
+            if (StringUtils.isBlank(user))
+            {
+                errors.reject(ERROR_MSG, "User name cannot be blank");
+            }
+            if (StringUtils.isBlank(password))
+            {
+                errors.reject(ERROR_MSG, "Password cannot be blank");
+            }
+            if (StringUtils.isBlank(doiPrefix))
+            {
+                errors.reject(ERROR_MSG, "DOI prefix cannot be blank");
+            }
+
+            String testUser = form.getTestUser();
+            String testPassword = form.getTestPassword();
+            String testDoiPrefix = form.getTestDoiPrefix();
+            if (StringUtils.isBlank(testUser))
+            {
+                errors.reject(ERROR_MSG, "Test user name cannot be blank");
+            }
+            if (StringUtils.isBlank(testPassword))
+            {
+                errors.reject(ERROR_MSG, "Test password cannot be blank");
+            }
+            if (StringUtils.isBlank(testDoiPrefix))
+            {
+                errors.reject(ERROR_MSG, "Test DOI prefix cannot be blank");
+            }
+        }
+
+        @Override
+        public boolean handlePost(DataCiteCredentialsForm form, BindException errors)
+        {
+            PropertyManager.PropertyMap map = PropertyManager.getEncryptedStore().getWritableProperties(DataCiteService.CREDENTIALS, true);
+            map.put(DataCiteService.USER, form.getUser());
+            map.put(DataCiteService.PASSWORD, form.getPassword());
+            map.put(DataCiteService.PREFIX, form.getDoiPrefix());
+            map.put(DataCiteService.TEST_USER, form.getTestUser());
+            map.put(DataCiteService.TEST_PASSWORD, form.getTestPassword());
+            map.put(DataCiteService.TEST_PREFIX, form.getTestDoiPrefix());
+            map.save();
+            return true;
+        }
+
+        @Override
+        public URLHelper getSuccessURL(DataCiteCredentialsForm form)
+        {
+            return null;
+        }
+
+        @Override
+        public ModelAndView getSuccessView(DataCiteCredentialsForm form)
+        {
+            ActionURL adminUrl = new ActionURL(JournalGroupsAdminViewAction.class, getContainer());
+            return new HtmlView(
+                    DIV("DataCite credentials saved!",
+                    BR(),
+                    new Link.LinkBuilder("Back to Panorama Public Admin Console").href(adminUrl).build()));
+        }
+
+        @Override
+        public ModelAndView getView(DataCiteCredentialsForm form, boolean reshow, BindException errors)
+        {
+            if(!reshow)
+            {
+                PropertyManager.PropertyMap map = PropertyManager.getEncryptedStore().getWritableProperties(DataCiteService.CREDENTIALS, false);
+                if(map != null)
+                {
+                    form.setUser(map.get(DataCiteService.USER));
+                    form.setPassword(map.get(DataCiteService.PASSWORD));
```

Comment:
Up to you, but we've gotten dinged on security reviews for echoing a previously entered password back into the HTML for forms like this. Instead, we've switched to forcing users to always reenter passwords when submitting the form.

Reply:
Done.